crates/rpc/rpc/src/eth/core.rs

```diff
@@ -181,6 +183,150 @@ where
+    N: RpcNodeCore,
+    EthApiError: reth_rpc_eth_api::FromEvmError<N::Evm>,
+    Rpc: RpcConvert<Primitives = N::Primitives, Error = EthApiError, Evm = N::Evm>,
+{
+    /// Returns the list of addresses that appeared in the given block.
+    ///
+    /// This includes addresses from:
+    /// - Block miner
+    /// - Transaction senders and recipients
+    /// - Withdrawals (post-Shanghai)
+    /// - Log emitter contracts
+    /// - Address-like values in log topics
+    /// - Accounts accessed during execution
+    pub async fn collect_addresses_in_block(
+        &self,
+        block_id: alloy_eips::BlockId,
+    ) -> Result<Vec<alloy_primitives::Address>, reth_rpc_eth_types::EthApiError> {
+        use alloy_consensus::{Transaction, TxReceipt};
+        use alloy_primitives::Address;
+        use reth_primitives_traits::BlockBody;
+        use reth_storage_api::{BlockIdReader, ReceiptProviderIdExt};
+        use std::collections::BTreeSet;
+
+        if block_id.is_pending() {
+            return Ok(Vec::new());
+        }
+
+        let block_hash = match self
+            .provider()
+            .block_hash_for_id(block_id)
+            .map_err(|e| reth_rpc_eth_types::EthApiError::Internal(e.into()))?
+        {
+            Some(hash) => hash,
+            None => return Ok(Vec::new()),
+        };
+
+        let block = match self
+            .cache()
+            .get_recovered_block(block_hash)
+            .await
+            .map_err(|e| reth_rpc_eth_types::EthApiError::Internal(e.into()))?
+        {
+            Some(block) => block,
+            None => return Ok(Vec::new()),
+        };
+
+        let mut set: BTreeSet<Address> = BTreeSet::new();
+
+        set.insert(block.header().beneficiary());
+        if let Some(withdrawals) = block.body().withdrawals() {
+            for withdrawal in withdrawals {
+                set.insert(withdrawal.address);
+            }
+        }
+
+        use reth_primitives_traits::SignerRecoverable;
+        for tx in block.body().transactions() {
+            if let Ok(signer) = tx.recover_signer() {
+                set.insert(signer);
+            }
+            if let Some(to) = tx.to() {
+                set.insert(Address::from(*to));
+            }
+            // Add addresses from access list (EIP-2930)
+            if let Some(access_list) = tx.access_list() {
+                for item in &access_list.0 {
+                    set.insert(item.address);
+                }
+            }
+        }
+
+        let this = self.clone();
+        let block_for_execution = block.clone();
+        let witness_record = self
+            .spawn_with_state_at_block(
+                block_for_execution.parent_hash().into(),
+                move |state_provider| {
+                    let db = StateProviderDatabase::new(&state_provider);
+                    let block_executor = this.evm_config().executor(db);
+
+                    let mut record = ExecutionWitnessRecord::default();
```

Comment:
I don't think we need this here and can instead take a look at the full state output and iterate over the accounts in there